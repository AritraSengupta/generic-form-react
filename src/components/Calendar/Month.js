import React from 'react'
import { Table } from 'semantic-ui-react'
import { getDateString, isValidDate } from '../../utils'

export const isDisabledCell = (date, month, year, startDisable, endDisable) => {
  if (!date) return true

  if (!isValidDate(startDisable) && !isValidDate(endDisable)) return false
  const start = isValidDate(startDisable) ? startDisable : -Infinity
  const end = isValidDate(endDisable) ? endDisable : Infinity
  const currentDate = new Date(year, month, date)

  return currentDate >= start && currentDate <= end
}

/**
 * type Props = {
 *  onChange: (e: Event, name: string) => void
 *  month: string
 *  year: string
 *  selectedOrCurrent: Date string (DD-MM-YYYY), showing date in blue, useful for showing selection
 *  highlightedDates: Array<{[date: number]: warning | negative | positive | error}>
 * }
 *
 * selectedOrCurrent is used to show a selected or current date when using a date picker
 * highlightedDates are to be used in calendar mode when each month is a different instance of the component
 */

export class MonthTable extends React.Component {
  _dayArray = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa']

  _monthArray = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec'
  ]

  getMonthData(month, year) {
    // Month + 1 because 0th date is last day of prev month
    const lastDate = new Date(year, month + 1, 0)
    if (isValidDate(lastDate)) {
      const days = lastDate.getDate()
      const day = lastDate.getDay()
      return this.createMonthObject([{}], day, days)
    }
    return null
  }

  _getClassName = (day, tableBodyProps, disabled) => {
    const date = tableBodyProps[day] || ''
    if (!date) return 'compact'
    if (disabled) return 'compact disabled'
    const { month, year, selectedOrCurrent } = this.props
    const cellDate = getDateString(date, month, year)
    return cellDate === selectedOrCurrent
      ? 'compact highlight selected'
      : 'compact highlight'
  }

  _clickDay = (e) => {
    const { onChange } = this.props
    onChange(e, e.target.dataset.name)
  }

  processCells = (tableBodyProps) => {
    const {
      highlightedDates,
      month,
      year,
      disableStart,
      disableEnd
    } = this.props
    const cells = []
    this._dayArray.forEach((day) => {
      const isDisabled = isDisabledCell(
        tableBodyProps[day],
        month,
        year,
        disableStart,
        disableEnd
      )
      const dayObj = {
        key: day,
        content: tableBodyProps[day] || '',
        className: this._getClassName(day, tableBodyProps, isDisabled),
        onClick: !isDisabled ? this._clickDay : null, // To prevent clicking outside a given date
        'data-name': tableBodyProps[day]
      }
      const specialSign =
        highlightedDates && highlightedDates[tableBodyProps[day]]
      if (specialSign) {
        dayObj[specialSign] = true
      }
      cells.push(dayObj)
    })
    return cells
  }

  getHeaderRow() {
    return {
      cells: this._dayArray.map((day) => ({
        key: day,
        style: { padding: 0 },
        content: day
      }))
    }
  }

  createMonthObject(data, dayIndex, date) {
    if (date === 0) return data

    const dayString = this._dayArray[dayIndex]

    // eslint-disable-next-line
    data[0][dayString] = date;

    if (dayIndex === 0 && date > 1) {
      data.unshift({})
      return this.createMonthObject(data, 6, date - 1)
    }

    return this.createMonthObject(data, dayIndex - 1, date - 1)
  }

  renderBodyRowNew = (tableBodyProps, i) => ({
    key: `row-${i}`,
    cells: this.processCells(tableBodyProps)
  })

  render() {
    const { month, year } = this.props
    const tableData = this.getMonthData(month, year)
    const headerRow = this.getHeaderRow()
    if (!tableData) {
      return <span> Invalid Date </span>
    }
    return (
      <Table
        size='small'
        textAlign='center'
        collapsing
        headerRow={headerRow}
        renderBodyRow={this.renderBodyRowNew}
        tableData={tableData}
        className='table-month'
      />
    )
  }
}

export default MonthTable
