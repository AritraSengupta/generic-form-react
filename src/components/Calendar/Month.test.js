import React from 'react'
import { fireEvent } from '@testing-library/react'
import Month, { isDisabledCell } from './Month'

describe('Month', () => {
  it('should render a invalid table if data is invalid', () => {
    const { container } = render(<Month month='April' year={null} />)
    expect(container).toMatchSnapshot()
    const tableObj = container.querySelectorAll('table')
    expect(tableObj.length).toBe(0)
  })
  it('should render a month table if valid date', () => {
    const { container } = render(
      <Month month={1} year={2019} disableStart='2019-2-15' />
    )
    expect(container).toMatchSnapshot()
    const tableObj = container.getElementsByTagName('table')
    expect(tableObj.length).toBe(1)
  })

  it('should be able to click and date', () => {
    const clickMock = jest.fn()
    const { container } = render(
      <Month month={1} year={2019} onChange={clickMock} />
    )
    const tableObj = container.querySelectorAll('td[data-name="16"]')
    expect(tableObj.length).toBe(1)
    fireEvent.click(tableObj[0])
    expect(clickMock).toHaveBeenCalledWith(expect.anything(), '16')
  })
})
