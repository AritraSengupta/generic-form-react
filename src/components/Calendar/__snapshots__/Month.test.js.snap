// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Month should render a invalid table if data is invalid 1`] = `
<div>
  <span>
     Invalid Date 
  </span>
</div>
`;

exports[`Month should render a month table if valid date 1`] = `
<div>
  <table
    class="ui small collapsing center aligned table table-month"
  >
    <thead
      class=""
    >
      <tr
        class=""
      >
        <th
          class=""
          style="padding: 0px;"
        >
          Su
        </th>
        <th
          class=""
          style="padding: 0px;"
        >
          Mo
        </th>
        <th
          class=""
          style="padding: 0px;"
        >
          Tu
        </th>
        <th
          class=""
          style="padding: 0px;"
        >
          We
        </th>
        <th
          class=""
          style="padding: 0px;"
        >
          Th
        </th>
        <th
          class=""
          style="padding: 0px;"
        >
          Fr
        </th>
        <th
          class=""
          style="padding: 0px;"
        >
          Sa
        </th>
      </tr>
    </thead>
    <tbody
      class=""
    >
      <tr
        class=""
      >
        <td
          class="compact"
        >
          
        </td>
        <td
          class="compact"
        >
          
        </td>
        <td
          class="compact"
        >
          
        </td>
        <td
          class="compact"
        >
          
        </td>
        <td
          class="compact"
        >
          
        </td>
        <td
          class="compact highlight"
          data-name="1"
        >
          1
        </td>
        <td
          class="compact highlight"
          data-name="2"
        >
          2
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          class="compact highlight"
          data-name="3"
        >
          3
        </td>
        <td
          class="compact highlight"
          data-name="4"
        >
          4
        </td>
        <td
          class="compact highlight"
          data-name="5"
        >
          5
        </td>
        <td
          class="compact highlight"
          data-name="6"
        >
          6
        </td>
        <td
          class="compact highlight"
          data-name="7"
        >
          7
        </td>
        <td
          class="compact highlight"
          data-name="8"
        >
          8
        </td>
        <td
          class="compact highlight"
          data-name="9"
        >
          9
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          class="compact highlight"
          data-name="10"
        >
          10
        </td>
        <td
          class="compact highlight"
          data-name="11"
        >
          11
        </td>
        <td
          class="compact highlight"
          data-name="12"
        >
          12
        </td>
        <td
          class="compact highlight"
          data-name="13"
        >
          13
        </td>
        <td
          class="compact highlight"
          data-name="14"
        >
          14
        </td>
        <td
          class="compact highlight"
          data-name="15"
        >
          15
        </td>
        <td
          class="compact highlight"
          data-name="16"
        >
          16
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          class="compact highlight"
          data-name="17"
        >
          17
        </td>
        <td
          class="compact highlight"
          data-name="18"
        >
          18
        </td>
        <td
          class="compact highlight"
          data-name="19"
        >
          19
        </td>
        <td
          class="compact highlight"
          data-name="20"
        >
          20
        </td>
        <td
          class="compact highlight"
          data-name="21"
        >
          21
        </td>
        <td
          class="compact highlight"
          data-name="22"
        >
          22
        </td>
        <td
          class="compact highlight"
          data-name="23"
        >
          23
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          class="compact highlight"
          data-name="24"
        >
          24
        </td>
        <td
          class="compact highlight"
          data-name="25"
        >
          25
        </td>
        <td
          class="compact highlight"
          data-name="26"
        >
          26
        </td>
        <td
          class="compact highlight"
          data-name="27"
        >
          27
        </td>
        <td
          class="compact highlight"
          data-name="28"
        >
          28
        </td>
        <td
          class="compact"
        >
          
        </td>
        <td
          class="compact"
        >
          
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
